src/core/sliding_window_extremum.c

```diff
@@ -0,0 +1,169 @@
+/*++
+
+    Copyright (c) Microsoft Corporation.
+    Licensed under the MIT License.
+
+Abstract:
+    The following functions implement a sliding window extremum (either maxima
+    or minima) algorithm for MsQuic. The design is based on a well-known data
+    structure called "monotone queue".
+
+    Since the queue is monotonic, we can easily access the extremum element by
+    looking at the first element. Adding new element will sweep out
+    older elements which value is greater/less than equal to the new element,
+    along with the elements which is expired.
+    
+    If there are more elements comparing to the capacity of queue, the
+    algorithm will still work but the accuracy will be compromised.
+
+--*/
+
+
+#include "precomp.h"
+
+_IRQL_requires_max_(DISPATCH_LEVEL)
+SLIDING_WINDOW_EXTREMUM
+SlidingWindowExtremumInitialize(
+    _In_ uint64_t EntryLifetime,
+    _In_ uint32_t WindowCapacity,
+    _In_ SLIDING_WINDOW_EXTREMUM_ENTRY* Window
+    )
+{
+    SLIDING_WINDOW_EXTREMUM w;
```

Comment:
Please, no single letter variables. This looks like it should be called `Window` and the input Window variable should be called `Entry`. Should `SLIDING_WINDOW_EXTREMUM->Window` be called `SLIDING_WINDOW_EXTREMUM->Entry`?

Reply:
> Please, no single letter variables.                

OK 

> Should SLIDING_WINDOW_EXTREMUM->Window be called SLIDING_WINDOW_EXTREMUM->Entry?

* `SLIDING_WINDOW_EXTREMUM->Entry` 
* `SLIDING_WINDOW_EXTREMUM->Entries`
* `SLIDING_WINDOW_EXTREMUM->Deque`
* `SLIDING_WINDOW_EXTREMUM->Extremums`

They were the candidates for the internal "window" of this data structure...

Reply:
I've changed the name of internal queue to `SLIDING_WINDOW_EXTREMUM->Extremums`.
Feel free to change if you have better options.